### PR TITLE
🐛 [Bug-Fix]: Mounted subapps don't work correctly if parent app attached …

### PR DESCRIPTION
**Pull request by ReneWerner87**
🐛 [Bug]: Mounted subapps don't work correctly if parent app attached additional middlewares after mounting (v2.40.1 bug) #2233

## Description
 
Pull request should fix the bug #2233 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [x] For new code I have written benchmarks so that they can be analyzed and improved

### CODE REVIEW

**Comment by ReneWerner87** [comment]
```ts
import express, { Request, Response } from 'express';

const app = express();

app.get('/hello', (req: Request, res: Response) => {
    res.send('Hello!');
});

const subApp = express();

subApp.get('/world', (req: Request, res: Response) => {
    res.send('World!');
});

app.use('/', subApp);

// 404 Handler
app.use((req: Request, res: Response) => {
    res.status(404).send('Not Found');
});

app.listen(3000, () => {
    console.log('Listening on port 3000');
});
```
```txt
http://localhost:3000/world -> Response: World!
http://localhost:3000/hello -> Response: Hello!
http://localhost:3000/test   -> Response: Not Found
```

**Comment by ReneWerner87** [comment]
```ts
import express, { Request, Response } from 'express';

const app = express();
const subApp = express();

subApp.get('/hello', (req: Request, res: Response) => res.send('Hello 2!'));
subApp.get('/world', (req: Request, res: Response) => res.send('World 2!'));

app.get('/hello', (req: Request, res: Response) => res.send('Hello 1!'));
app.use('/', subApp);
app.get('/world', (req: Request, res: Response) => res.send('World 1!'));

// 404 Handler
app.use((req: Request, res: Response) => {
    res.status(404).send('Not Found');
});

app.listen(3000, () => {
    console.log('Listening on port 3000');
});
```

```txt
http://localhost:3000/world -> Response: World 2!
http://localhost:3000/hello -> Response: Hello 1!
http://localhost:3000/test  -> Response: Not Found
```

File: helpers.go [hunk 1]
```diff
@@ -93,19 +93,19 @@ func (app *App) methodExist(ctx *Ctx) bool {
 			continue
 		}
 		// Reset stack index
-		ctx.indexRoute = -1
+		indexRoute := -1
```
Comment (ReneWerner87) [inline helpers.go:96]: check the allocations later

**Comment by ReneWerner87** [comment]
image
has anyone an idea about the reason for this ?

**Comment by li-jin-gou** [comment]
> image has anyone an idea about the reason for this ?

I currently suspect that the github action is unstable and causing this problem.

**Comment by gaby** [comment]
@ReneWerner87 I'm honestly not fully up-to-speed on what this code does. Using GPT4 I was able to get some suggestion that may be useful:

### Implementation suggestions for  `processSubAppsRoutes()` :
- Avoid slice re-allocation: When inserting sub-app routes into the parent app's stack, you are using append multiple times, which can lead to slice re-allocation. Instead, preallocate space for the new slice and copy the necessary elements directly.

```go
newStack := make([]*Route, len(app.stack[m]) + len(subRoutes) - 1)
copy(newStack[:i], app.stack[m][:i])
copy(newStack[i:i+len(subRoutes)], subRoutes)
copy(newStack[i+len(subRoutes):], app.stack[m][i+1:])
app.stack[m] = newStack
```

- Use a separate variable for the atomic operation: The atomic operations can be replaced with a local variable that is updated inside the loop, and then stored atomically at the end of the function.

```go
// Replace atomic.AddUint32(&handlersCount, uint32(len(route.Handlers))) with:
handlersCount += uint32(len(route.Handlers))

// At the end of the function:
atomic.StoreUint32(&app.handlersCount, handlersCount)
```

- Mutex usage: The `sync.Once` is used to ensure that certain parts of the code are executed only once. While this approach is effective, it might be worth exploring other synchronization mechanisms like `sync.Mutex` to provide better control over concurrency, especially in high-load scenarios.

### The following were given regarding other functions and their implementation:

- The provided Mount function can be optimized in a few ways. Here's the modified version with optimizations:

```go
func (app *App) Mount(prefix string, subApp *App) Router {
    prefix = strings.TrimRight(prefix, "/")
    if prefix == "" {
        prefix = "/"
    }

    // Optimize: Pre-allocate space for app.mountFields.appList map
    n := len(subApp.mountFields.appList)
    if n > 0 && app.mountFields.appList == nil {
        app.mountFields.appList = make(map[string]*App, n)
    }

    // Support for configs of mounted-apps and sub-mounted-apps
    for mountedPrefixes, subApp := range subApp.mountFields.appList {
        path := getGroupPath(prefix, mountedPrefixes)

        subApp.mountFields.mountPath = path
        app.mountFields.appList[path] = subApp
    }

    // register mounted group
    mountGroup := &Group{Prefix: prefix, app: subApp}
    app.register(methodUse, prefix, mountGroup)

    // Execute onMount hooks
    if err := subApp.hooks.executeOnMountHooks(app); err != nil {
        panic(err)
    }

    return app
}
```

- The same applies for `func (grp *Group) Mount(prefix string, subApp *App) Router` by using:

```go
n := len(subApp.mountFields.appList)
if n > 0 && grp.app.mountFields.appList == nil {
    grp.app.mountFields.appList = make(map[string]*App, n)
}
```

- The provided `appendSubAppLists` function can be optimized in the following ways:

Pre-allocate space for app.mountFields.appList map.
Avoid calling getGroupPath() function multiple times in case it's an expensive operation.
Here's the optimized version:

```go
func (app *App) appendSubAppLists(appList map[string]*App, parent ...string) {
	// Optimize: Pre-allocate space for app.mountFields.appList map
	n := len(appList)
	if n > 0 && app.mountFields.appList == nil {
		app.mountFields.appList = make(map[string]*App, n)
	}

	// Optimize: Cache parent prefix
	parentPrefix := ""
	if len(parent) > 0 {
		parentPrefix = parent[0]
	}

	for prefix, subApp := range appList {
		// skip real app
		if prefix == "" {
			continue
		}

		if parentPrefix != "" {
			prefix = getGroupPath(parentPrefix, prefix)
		}

		if _, ok := app.mountFields.appList[prefix]; !ok {
			app.mountFields.appList[prefix] = subApp
		}

		// The first element of appList is always the app itself. If there are no other sub apps, we should skip appending nested apps.
		if len(subApp.mountFields.appList) > 1 {
			app.appendSubAppLists(subApp.mountFields.appList, prefix)
		}
	}
}
```

Optimizations applied:

- Pre-allocate space for app.mountFields.appList map: If app.mountFields.appList is nil and appList has elements, pre-allocate space for app.mountFields.appList map to avoid reallocations when adding entries to it.
```go
n := len(appList)
if n > 0 && app.mountFields.appList == nil {
	app.mountFields.appList = make(map[string]*App, n)
}
```

- Cache parent prefix: Avoid calling getGroupPath() multiple times by caching parentPrefix. This may improve performance if getGroupPath() is an expensive operation.
```go
parentPrefix := ""
if len(parent) > 0 {
	parentPrefix = parent[0]
}
```

**Comment by ReneWerner87** [comment]
> ### Implementation suggestions for `processSubAppsRoutes()` :
> * Avoid slice re-allocation: When inserting sub-app routes into the parent app's stack, you are using append multiple times, which can lead to slice re-allocation. Instead, preallocate space for the new slice and copy the necessary elements directly.
> 
> ```go
> newStack := make([]*Route, len(app.stack[m]) + len(subRoutes) - 1)
> copy(newStack[:i], app.stack[m][:i])
> copy(newStack[i:i+len(subRoutes)], subRoutes)
> copy(newStack[i+len(subRoutes):], app.stack[m][i+1:])
> app.stack[m] = newStack
> ```
done ✅
> * Use a separate variable for the atomic operation: The atomic operations can be replaced with a local variable that is updated inside the loop, and then stored atomically at the end of the function.
> 
> ```go
> // Replace atomic.AddUint32(&handlersCount, uint32(len(route.Handlers))) with:
> handlersCount += uint32(len(route.Handlers))
> 
> // At the end of the function:
> atomic.StoreUint32(&app.handlersCount, handlersCount)
> ```
done ✅
> * Mutex usage: The `sync.Once` is used to ensure that certain parts of the code are executed only once. While this approach is effective, it might be worth exploring other synchronization mechanisms like `sync.Mutex` to provide better control over concurrency, especially in high-load scenarios.

☑️ not required, as it is a process that does not run in parallel
and at the start of the application

> ### The following were given regarding other functions and their implementation:
> * The provided Mount function can be optimized in a few ways. Here's the modified version with optimizations:
> 
> ```go
> func (app *App) Mount(prefix string, subApp *App) Router {
>     prefix = strings.TrimRight(prefix, "/")
>     if prefix == "" {
>         prefix = "/"
>     }
> 
>     // Optimize: Pre-allocate space for app.mountFields.appList map
>     n := len(subApp.mountFields.appList)
>     if n > 0 && app.mountFields.appList == nil {
>         app.mountFields.appList = make(map[string]*App, n)
>     }
> 
>     // Support for configs of mounted-apps and sub-mounted-apps
>     for mountedPrefixes, subApp := range subApp.mountFields.appList {
>         path := getGroupPath(prefix, mountedPrefixes)
> 
>         subApp.mountFields.mountPath = path
>         app.mountFields.appList[path] = subApp
>     }
> 
>     // register mounted group
>     mountGroup := &Group{Prefix: prefix, app: subApp}
>     app.register(methodUse, prefix, mountGroup)
> 
>     // Execute onMount hooks
>     if err := subApp.hooks.executeOnMountHooks(app); err != nil {
>         panic(err)
>     }
> 
>     return app
> }
> ```
> 
> * The same applies for `func (grp *Group) Mount(prefix string, subApp *App) Router` by using:
> 
> ```go
> n := len(subApp.mountFields.appList)
> if n > 0 && grp.app.mountFields.appList == nil {
>     grp.app.mountFields.appList = make(map[string]*App, n)
> }
> ```

☑️ not needed, because it happens before
see:
https://github.com/gofiber/fiber/blob/2237e9c5115e6ade2bfedf6a85d04a36957cd635/app.go#L512
https://github.com/gofiber/fiber/blob/2237e9c5115e6ade2bfedf6a85d04a36957cd635/mount.go#L27-L32


> * The provided `appendSubAppLists` function can be optimized in the following ways:
> 
> Pre-allocate space for app.mountFields.appList map. Avoid calling getGroupPath() function multiple times in case it's an expensive operation. Here's the optimized version:
> 
> ```go
> func (app *App) appendSubAppLists(appList map[string]*App, parent ...string) {
> 	// Optimize: Pre-allocate space for app.mountFields.appList map
> 	n := len(appList)
> 	if n > 0 && app.mountFields.appList == nil {
> 		app.mountFields.appList = make(map[string]*App, n)
> 	}
> 
> 	// Optimize: Cache parent prefix
> 	parentPrefix := ""
> 	if len(parent) > 0 {
> 		parentPrefix = parent[0]
> 	}
> 
> 	for prefix, subApp := range appList {
> 		// skip real app
> 		if prefix == "" {
> 			continue
> 		}
> 
> 		if parentPrefix != "" {
> 			prefix = getGroupPath(parentPrefix, prefix)
> 		}
> 
> 		if _, ok := app.mountFields.appList[prefix]; !ok {
> 			app.mountFields.appList[prefix] = subApp
> 		}
> 
> 		// The first element of appList is always the app itself. If there are no other sub apps, we should skip appending nested apps.
> 		if len(subApp.mountFields.appList) > 1 {
> 			app.appendSubAppLists(subApp.mountFields.appList, prefix)
> 		}
> 	}
> }
> ```
> 
> Optimizations applied:
> 
> * Pre-allocate space for app.mountFields.appList map: If app.mountFields.appList is nil and appList has elements, pre-allocate space for app.mountFields.appList map to avoid reallocations when adding entries to it.
> 
> ```go
> n := len(appList)
> if n > 0 && app.mountFields.appList == nil {
> 	app.mountFields.appList = make(map[string]*App, n)
> }
> ```
> 
> * Cache parent prefix: Avoid calling getGroupPath() multiple times by caching parentPrefix. This may improve performance if getGroupPath() is an expensive operation.
> 
> ```go
> parentPrefix := ""
> if len(parent) > 0 {
> 	parentPrefix = parent[0]
> }
> ```

✅ done, without the make because of the mentioned process before

**Review by gaby — APPROVED** [review]

**Pull request merged**